File: src/MphRead.Native/Mods/Platform/AppPaths.cpp
#include "AppPaths.hpp"

#include "../Branding.hpp"
#include "../../NativeRuntime/System/IO.hpp"
#include "../../NativeRuntime/System/Runtime.hpp"

#include <memory>

namespace MphRead::Mods::Platform
{
    std::string AppPaths::ExecutableDirectory()
    {
        return ::MphRead::NativeRuntime::AppContextBaseDirectory();
    }

    std::string AppPaths::ResourceDirectory()
    {
        const ::MphRead::NativeRuntime::DirectoryInfo executable(ExecutableDirectory());
        const std::shared_ptr<::MphRead::NativeRuntime::DirectoryInfo> parent = executable.Parent();
        const std::shared_ptr<::MphRead::NativeRuntime::DirectoryInfo> grandparent
            = parent == nullptr ? nullptr : parent->Parent();
        const bool bundled = ::MphRead::NativeRuntime::IsMacOS() && executable.Name() == "MacOS"
            && parent != nullptr && parent->Name() == "Contents"
            && grandparent != nullptr && grandparent->Extension() == ".app";
        return bundled
            ? ::MphRead::NativeRuntime::PathCombine(parent->FullName(), "Resources")
            : ExecutableDirectory();
    }

    std::string AppPaths::Maps()
    {
        return ::MphRead::NativeRuntime::PathCombine(ResourceDirectory(), "maps");
    }

    std::string AppPaths::UserDataDirectory()
    {
        return ::MphRead::NativeRuntime::IsMacOS()
            ? ::MphRead::NativeRuntime::PathCombine(
                ::MphRead::NativeRuntime::EnvironmentUserProfile(),
                "Library", "Application Support", std::string(Branding::Name))
            : ExecutableDirectory();
    }

    std::string AppPaths::PathsFile()
    {
        return ::MphRead::NativeRuntime::PathCombine(UserDataDirectory(), "paths.txt");
    }

    void AppPaths::PrepareUserData()
    {
        if (::MphRead::NativeRuntime::IsMacOS())
        {
            ::MphRead::NativeRuntime::DirectoryCreateDirectory(UserDataDirectory());
        }
    }
}

File: src/MphRead/Mods/Platform/AppPaths.cs
using System;
using System.IO;

namespace MphRead.Mods.Platform
{
    /// <summary>Installation resources and writable desktop state have different roots.</summary>
    internal static class AppPaths
    {
        public static string ExecutableDirectory => AppContext.BaseDirectory;
        public static string ResourceDirectory
        {
            get
            {
                var executable = new DirectoryInfo(ExecutableDirectory);
                bool bundled = OperatingSystem.IsMacOS() && executable.Name == "MacOS"
                    && executable.Parent?.Name == "Contents"
                    && executable.Parent.Parent?.Extension == ".app";
                return bundled ? Path.Combine(executable.Parent!.FullName, "Resources") : ExecutableDirectory;
            }
        }
        public static string Maps => Path.Combine(ResourceDirectory, "maps");

        // Other desktop platforms keep their existing portable layout. Android
        // sets GameFiles.Root and LauncherPrefs.Directory from its activity.
        public static string UserDataDirectory => OperatingSystem.IsMacOS()
            ? Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
                "Library", "Application Support", Branding.Name)
            : ExecutableDirectory;

        public static string PathsFile => Path.Combine(UserDataDirectory, "paths.txt");

        public static void PrepareUserData()
        {
            if (OperatingSystem.IsMacOS())
            {
                Directory.CreateDirectory(UserDataDirectory);
            }
        }
    }
}
